### PR TITLE
Image loading via `loadMovie` et all

### PR DESCRIPTION
**Pull request by kmeisthax**
This PR enables loading images wherever you would ordinarily load an SWF into a clip. We content-sniff the headers to determine if the loaded movie is an SWF or an image; and if it's the latter we package it into a `Bitmap` and insert it into the target movie clip.

TODO:

 * [x] Write tests
 * ~~Determine if we need to *replace* target clips with `Bitmap`s instead of just loading the image into them; and if so, figure out how to do that~~
   * We don't
 * [x] Determine what happens if you load an image into the root movie
   * It clears the target level but fails to load into the movie

### CODE REVIEW

**Comment by adrian17** [comment]
General code comment: can we split `movie_loader` a bit, say the MC path from bitmap path, from error path? It's gotten really long. If it's not that easy because of async closure, maybe at least move some common fragments out? For example,
```rs
                    Avm1::run_stack_frame_for_method(
                        clip,
                        broadcaster,
                        NEWEST_PLAYER_VERSION,
                        uc,
                        "broadcastMessage".into(),
                        &["onLoadStart".into(), clip.object()],
                    );
```
Looks like a big boilerplate-y chunk repeated with minor variations.

(side question unrelated to PR: why is `broadcastMessage` called normally in all places except in `movie_clip_loaded`, where it's queued?)

**Comment by adrian17** [comment]
In test `mistarget`, what does the name stand for? :sweat_smile:

File: core/src/loader.rs [hunk 1]
```diff
@@ -4,21 +4,129 @@ use crate::avm1::activation::{Activation, ActivationIdentifier};
...
+        } else {
+            Err(Error::UnexpectedData(ExpectedType::Any, self))
+        }
+    }
+}
+
+/// A ContentType we expected to get, but didn't.
+///
+/// Exists purely because we can't use `Option` with `thiserror`. The
+/// alternative was `expect_known` erroring with something like "Expected
```
Comment (adrian17) [inline core/src/loader.rs:93]: Alternatively, add a completely new error variant for "couldn't recognize loaded file type"; this way, I'd argue that we might not need `ExpectedType` enum to exist at all.

**Comment by kmeisthax** [comment]
`mistarget` is just "mistarget", because the test loads a PNG that's intentionally mislabeled as a GIF. It's to check that content sniffing is being used over other file type signals.

**Comment by XXLuigiMario** [comment]
Beautiful rendition of the Ruffle logo 10/10

File: core/src/loader.rs [hunk 2]
```diff
@@ -410,46 +487,82 @@ impl<'gc> Loader<'gc> {
...
+                                    uc.library
+                                        .library_for_movie_mut(movie.clone())
+                                        .register_character(
+                                            id,
+                                            crate::character::Character::MorphShape(morph_shape),
+                                        );
+                                }
+                            }
                         }
+                        ContentType::Gif | ContentType::Jpeg | ContentType::Png => {
```
Comment (Herschel) [inline core/src/loader.rs:540]: TODO for future PR: I believe this is actually version-gated, v7 SWFs can not load PNGs, for example. Not sure about the specifics, but worth a comment.

**Comment by Herschel** [comment]
Any idea if MIME type affects how the official Flash plugin behaves when loading assets over HTTP? Might be worth a test.

File: core/src/loader.rs [hunk 3]
```diff
@@ -410,46 +487,82 @@ impl<'gc> Loader<'gc> {
...
+                                        clip.object(),
+                                        length.into(),
+                                        length.into(),
+                                    ],
+                                );
+                            }
+
+                            if let Some(mut mc) = clip.as_movie_clip() {
+                                mc.insert_at_index(uc, bitmap_obj.into(), 0);
+                            }
```
[the remark this answers is not in the document]
Comment (kmeisthax) [reply core/src/loader.rs:516]: Does it always use depth 1 or does it use `getNextHighestDepth`?

File: core/src/loader.rs (hunk 2 above)
Comment (kmeisthax) [reply core/src/loader.rs:498]: The exact specifics are that only `ContentType::Jpeg` works on v7. Unfortunately I can't build tests for it as the output is identical in both cases; one just doesn't render the image.

I have to wonder if this also applies to `DefineBitsJpeg2` tags in SWFs.

**Comment by kmeisthax** [comment]
Flash ignores the HTTP MIME, too:

image

(Ignore the domain name, I had a hostsfile override for it already set up for testing domain locks and it was convenient enough to get Flash content into)

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**